### DISCUSSION TITLE
Improving page load speed

### DISCUSSION THREAD

**Discussion by scottcwilson** [zencart/zencart#5861] [Ideas]
From https://github.com/zencart/zencart/pull/5845#issuecomment-1620926040: 

>> Anybody know why we're supporting self-hosting of the font-awesome, bootstrap and jquery payloads?

> Self hosting has benefits in SEO, because you can reduce the number of resources fetched for a page by consolidating them into single files. I know the current ZC approach doesn't do this, but in our shop for example we aggressively compress up to 8 CSS files into one minified one so our shop pages only fetch a single CSS file that's customised to the page being loaded. 

I have also experimented with this, and the results are quite striking. @neekfenwick I wonder if this is something we could do in a future release.

**Comment by drbyte** [comment]
Historically the template system provided "full" versions of 3rd party packages (eg: jQuery, etc), so that template designers and plugin creators could depend on all its features being present, instead of themselves injecting a full copy of the package, that got outdated fast, in order to make things work. 
Yes, the downside was size.

And yes, the latest trend is to self-minify everything with something like webpack and all its latest siblings. Don't expect the average storeowner to have any understanding of how to do that. 
But for sure, support its use if someone wants to do that.

**Comment by neekfenwick** [comment]
@scottcwilson to be clear, this isn't so much about 'page construction time' as 'page load speed'.  I've spent a lot of time with pagespeed.dev improving our 'First Contentful Paint', 'Speed Index' etc.  One big hit was the number of resources being loaded after initial page load: CSS files, JS files etc.

@drbyte yeah users aren't going to like Webpack.  It seems ZC still strives to be an out-of-the-box solution with no pre-deploy build steps involved, not even a 'composer install' or running of package.json scripts.  I do see why, but it does limit us a great deal.

It is possible we could impose a CSS loading mechanism in ZC where Core files and addons could register a CSS file as being "to be loaded", then as the header is output those files are identified, hashed for changes, and a single CSS file built if out of date, and served to contain all those files?  I'm currently using a Java tool [css-compressor](https://github.com/samilyak/css-compressor) and it does involve replacing some paths so e.g. webfonts still load OK, but we would not want a Java dependency in Core.  Correct me if I'm missing something obvious like a Laravel or other framework that already does this in a way we could use. The steps would be something like:

- Core and addons contribute to a global array, obviously we would abstract away the management behind an class/API, i.e. `$cssManager->addCSSInclude('includes/templates/my_template/css/funkystuff.css')`
- Core header runs a `$cssManager->generateCSSIncludes()` which:
  - Identifies if any CSS includes have changed since the last run, if so...
    - Load all CSS files, minify them, replace relative paths, and output to a file with a unique id, similar to webpack's hash approach
  - Emit a single `<link rel="stylesheet" href="compressed-css-ab123.css">`

Not reinventing anything here, it's just bundle/layer generation like webpack and friends have done for years, but would be done on-the-fly by ZC code instead of as a pre-deploy build step.

This doesn't touch on JS minification which is probably beyond our scope.  I use Google Closure Compiler via the Dojo build system (removed jQuery from our ZC), not sure how we could incorporate something like that in ZC, and anyway ZC only uses jQuery which is already minified.

**Comment by scottcwilson** [comment]
Title updated for clarity.
